addons/yatl/event/event.gd

```diff
@@ -43,3 +44,190 @@ const StreamOfflineEvent: Resource = preload("./types/events.gd").StreamOfflineE
 const UserAuthorizationGrantEvent: Resource = preload("./types/events.gd").UserAuthorizationGrantEvent
 const UserAuthorizationRevokeEvent: Resource = preload("./types/events.gd").UserAuthorizationRevokeEvent
 const UserUpdateEvent: Resource = preload("./types/events.gd").UserUpdateEvent
+
+
+# Private constants
+
+const __EVENT_WSS_URL: String = "ws://127.0.0.1:25708"
+
+
+# Private variables
+
+var __pal: __PAL
+
+var __broardcaster_user_id: String
+var __client_id: String
+var __access_token: String
+
+var __connection: __WebSocketConnection
+
+var __websocket_id: String
+var __websocket_frequency_min: float
+
+var __time_since_last_message: float = 0.0
+
+var __connections: Dictionary = {
+	# key: String - event
+	# value: [[FuncRef, Array - binds], ...]
+}
+
+
+# Lifecycle methods
+
+func _init(
+	_pal: __PAL,
+	_broadcaster_user_id: String,
+	_client_id: String,
+	_access_token: String
+) -> void:
+	__pal = _pal
+	__broardcaster_user_id = _broadcaster_user_id
+	__client_id = _client_id
+	__access_token = _access_token
+
+
+func _process(delta: float) -> void:
+	if !__connection:
+		return
+
+	__time_since_last_message += delta
+
+
+# Pubic methods
+
+func connect_event(
+	_event: String,
+	_target: Object,
+	_method: String,
+	_binds: Array = []
+) -> int:
+	# TODO: Check if event is a valid event
+	# return 1
+
+	if !__connection && !yield(__establish_connection(__EVENT_WSS_URL), "completed"):
+		return 2 # Some kind of error
+
+	if __connections.has(_event):
+		__connections[_event].append([
+			funcref(_target, _method),
+			_binds,
+		])
+
+		return OK
+
+	# TODO: Convert this to use API method
+	var response = __pal.request(
+		'http://127.0.0.1:25708/helix/eventsub/subscriptions',
+		{
+			"client-id": __client_id,
+			"authorization": "Bearer %s" % __access_token,
+			"content-type": "application/json",
+		},
+		false,
+		HTTPClient.METHOD_POST,
+		to_json({
+			"type": _event,
+			"version": "1",
+			"condition": {
+				"broadcaster_user_id": __broardcaster_user_id,
+			},
+			"transport": {
+				"method": "websocket",
+				"id": __websocket_id,
+			},
+		})
+	)
+
+	response = yield(response, "completed")
+
+	if response.error:
+		return 3
+
+	if response.response_code != 200:
+		print(response.response_code)
+		print(response.body)
+		return 4
+
+	__connections[_event] = []
+	__connections[_event].append([
+		funcref(_target, _method),
+		_binds,
+	])
+
+	return OK
+
+
+# Private methods
+
+func __establish_connection(url: String) -> bool:
+	__connection = yield(__pal.establish_connection(url), "completed")
+
+	# TODO: Determine if reconnect should be tried
```

Comment:
What does that mean?

Reply:
We should define the logic for retrying the connection if it wasn't able to connect in the first place. 
- How many retries
- Exponential back off
- etc

Reply:
Good idea. Since it would also be used in the Chat class, we should probably look at making that functionality present in WebsocketConnection?